[STEP-9] 주문완료 정보 카프카 메시지로 발행

n3rno opened the pull request:
## PR 설명
- 주문완료 정보 외부 전송 기능을 kafka 활용하도록 변경


## Definition of Done (DoD)
- [x] 카프카에 대한 기초 개념 문서로 정리 `kafka.md`
- [x] 주문 완료 이벤트 카프카 메시지로 발행하도록 변경
- [x] redis 이용하여 쿠폰 발급 기능 추가

n3rno committed:
카프카에 대한 기초 개념 및 구성요소 문서로 정리

n3rno committed:
스프링부트 카프카 연동 test

n3rno committed:
주문 완료 이벤트를 카프카 메시지로 발행하도록 변경

TODO TEST

n3rno committed:
orderService 버전 정리

n3rno committed:
REDIS 이용하여 쿠폰 발급 기능 추가

Isaac482763 commented on src/main/java/kr/hhplus/be/server/coupon/application/service/CouponIssueService.java:
```diff
@@ -0,0 +1,89 @@
+        }
+
+        // 수량 증가 후 최대 수량 초과 확인
+        Long currentCount = redisTemplate.opsForValue().increment(countKey);
+        int maxCount = couponRepository.findById(requestDto.getCouponId())
+                .map(Coupon::getMaxQuantity)
+                .orElseThrow(() -> new EcommerceException(ErrorCode.NOT_ENOUGH_COUPON));
```
maxCount를 먼저 조회하고 currentCount < maxCount 인 겨우에만 increment를 수행하도록 수정하는게 좋을것같습니다.

Isaac482763 commented on src/main/java/kr/hhplus/be/server/coupon/application/service/CouponIssueService.java:
```diff
@@ -0,0 +1,89 @@
+        if (Boolean.FALSE.equals(redisTemplate.hasKey(countKey))) {
+            Coupon coupon = couponRepository.findById(requestDto.getCouponId())
+                    .orElseThrow(() -> new EcommerceException(ErrorCode.NOT_EXIST_COUPON));
+            redisTemplate.opsForValue().set(countKey, "0");
```
여러 사용자가 동시에 호출하는 경우
동시에 초기화되어 increment()로 증가된값이 다시 0으로 초기화될수 있습니다.
setIfAbsent() 또는 Lua 스크립트를 활용해 원자적으로 처리하도록 하는것이 어떨까합니다.

setIfAbsent 예시
```
Boolean isNew = redisTemplate.opsForValue().setIfAbsent(countKey, "0", Duration.ofHours(1));
if (Boolean.TRUE.equals(isNew)) {
    // countKey가 없어서 새로 설정함 
}
```

Lua script 예시
```
if redis.call('EXISTS', KEYS[1]) == 0 then
  redis.call('SET', KEYS[1], ARGV[1])
  redis.call('EXPIRE', KEYS[1], ARGV[2])
end
```

Isaac482763 commented on src/main/java/kr/hhplus/be/server/kafka/service/KafkaConsumerService.java:
```diff
@@ -0,0 +1,20 @@
+    // [컨슈머] 메시지 수신
+    @KafkaListener(topics = "test-topic", groupId = "consumer-group-a")
+    public void listen(ConsumerRecord<String, String> record) {
+        System.out.println("Received: " + record.value());
```
카프카로 수신한 string 값을 역직렬화하여 객체로 변환하는 코드를 추가해보면 좋을것같습니다.

Isaac482763 replied:
이미 처리된 이벤트인지 확인하는 코드가 추가되면 좋을거같아요 ㅎㅎ
```
              if (processedEventRepository.isProcessed(event.getEventId())) { 
                 System.out.println("이미 처리된 이벤트입니다: " + event.getEventId());
                 return;
             }
```

n3rno committed:
코드리뷰 수정 반영

- 쿠폰발행 시 validation 논리 로직 순서 변경
- 쿠폰 발행 redis setIfAbsent 사용하여, key가 존재하지 않는 경우만 value를 설정한다.
- kafka로 수신한 string 값을 역직렬화하여 객체로 변환한다.
